solutions/gold/ys-MaxIndepSet.mdx

```diff
@@ -2,216 +2,168 @@
+Then it remains to find the maximum [clique](https://en.wikipedia.org/wiki/Clique_(graph_theory))
+in the graph instead of the maximum set.
+Note that this doesn't change anything about the actual problem- this is just to simplify things a little.
 
-But the time complexity is $\mathcal{O}(2 ^ n)$ or $\mathcal{O}(2 ^ n \cdot n)$,
-so we'll need to optimize it further.
+First, let's split the graph in two as previously stated in the hints.
+For all node subsets of each half, we then calculate:
+1. Whether the subset *itself* is a clique
+2. The maximum clique that is a subset of the node subset
 
-#### Meet-in-the-middle
+Then, when merging the halves, we iterate through all cliques of the first half of the graph
+and check which nodes in the other half are connected to all nodes in the first clique.
+With these nodes in the other half, we can then look at the maximum clique we previously calculated
+and add them to the clique in the first half.
 
-Divide the original graph into two parts, do DP on each part, then merge:
-
-Given two parts $S1$ and $S2$, loop through **every subgraph** of $S1$,
-
-Then, find a subgraph of $S2$ that has direct edges with **all of the points** in the subgraph of $S1$ that we're looping through.
-
-Lastly, we'll union the two subgraphs.
-
-Finding this subgraph of $S2$ can also be solved with Bitmask DP.
-
-Time Complexity is $\mathcal{O}(2^{\frac N2})$.
-</Spoiler>
-
-<Spoiler title="Implementation">
+## Implementation
 
 **Time Complexity:** $\mathcal{O}(2^{\frac N2})$
 
 <LanguageSection>
-
 <CPPSection>
 
 ```cpp
-#include <cstdio>
+#include <bitset>
+#include <cassert>
 #include <iostream>
 #include <vector>
-using namespace std;
+
+using std::cout;
+using std::endl;
+using std::vector;
 using ll = long long;
 
-// this function returns the number with more 1-bits (in binary)
-ll popmax(ll x, ll y) {
-	/*
-	 * __builtin_popcountll(x) returns the number of 1-bits of unsigned long
-	 * long x (in binary)
-	 */
-	return __builtin_popcountll(x) < __builtin_popcountll(y) ? y : x;
+/** @return whether n has a bit set at the position given */
+bool bit_set(const ll &n, int pos) { return n & (1LL << pos); }
+
+/** @return the number given with more bits set */
+ll more_bits(const ll &a, const ll &b) {
+	return __builtin_popcountll(a) > __builtin_popcountll(b) ? a : b;
 }
 
-void maxclique(int x, int dt, vector<ll> &f, vector<ll> a) {
-	for (int i = 0, c, d; i < (1 << x); i++) {
-		d = i & (-i);
-		/*
-		 * __builtin_ctzll(d) counts the trailing number of 0-bits of unsigned
-		 * long long d (in binary)
-		 */
-		c = __builtin_ctzll(d);
-		f[i] = popmax(f[i ^ d], f[i & a[c + dt] >> (dt)] | d);
+/** @return a pair with the two things our subsets keep track of */
+std::pair<vector<bool>, vector<ll>> max_cliques(const vector<ll> &adj) {
+	int sz = adj.size();  // just a shorthand
+
+	// This keeps track of whether subsets are cliques themselves
+	vector<bool> valid(1 << sz);
```

Comment:
would bitset work better than vector<bool> here?

Reply:
i mean i'd rather dynamically size it

Reply:
ah I see